File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.js
function solution(numbers, target) {
    var answer = 0;
    let arr = [];
    for(let n of numbers) arr.push(n);
    let visited = new Array(numbers.length).fill(false);
    let select = [];
    
    function dfs(depth){
        if(depth === numbers.length){
            let s = "";
            for(let i = 0; i < numbers.length; i++){
                s += select[i] + arr[i];
            }
            let result = eval(s);
            if(result === target) answer++;
            return;
        }
        for(let x of ["+" , "-"]){
            select.push(x);
            dfs(depth + 1);
            select.pop();
        }
    }
    dfs(0);
    
    return answer;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    
    answer = 0
    def dfs(idx , temp):
        nonlocal answer
        if idx == len(numbers):
            if temp == target:
                answer += 1
                return True
            return False
        
        # 음수일 경우
        dfs(idx + 1 , temp - numbers[idx])
            
        # 양수일 경우
        dfs(idx + 1 , temp + numbers[idx])
        
    dfs(0 , 0)
    
    return answer